Profile details page integration

MilesDev91 opened the pull request:
### What this PR does:
- Hooks the profile details page to the backend.
- Profile is retrieved based on id passed in params /profile/:id
- Requests can be sent via the request form.

### Screenshots / Video:
![image](https://user-images.githubusercontent.com/34494229/150627356-ac20ac92-b085-4714-ae5c-637ffd2c930d.png)
![image](https://user-images.githubusercontent.com/34494229/150627424-f9984a79-7e9f-4d62-98f7-b247589e6cc8.png)
![image](https://user-images.githubusercontent.com/34494229/150627475-f630b433-cdf9-4425-ab7f-7d8256725fb9.png)


### Any information needed to test this feature:
- Sign up as a new user to populate the necessary fields. Some can be edited in the profile settings page, others are not implemented and must be changed directly in the database or left as empty strings.
- Make sure you have a valid Profile to retrieve, in it you will want: name, photo (url), location, description, aboutMe, isSitter, and payRate. Photo is the most optional one (as it has a placeholder), the others will just default to empty strings. isSitter must be true or the page won't load the profile at all.

- Enter in a valid date range into the request form and submit. You should see a message that it was successful. Check your server to ensure successful creation.

### Any issues with the current functionality:
- The extra images are not currently implemented in the model, thus mock images will be used until they are there and uploadable. It should be a very easy integration once it is done.
- Not all values can be set in the settings, so naturally creating and editing a profile will not show all available information (although you can still send requests).
- Natural navigation to the profile details is not yet implemented. We currently must enter the profile id manually into the url.

closes #35

MilesDev91 committed:
Merge branch 'manage-bookings-integration' into profile-details-page-integration

MilesDev91 committed:
API setup

Set up the profile interface, API call, method, and route for fetching a single profile.

Added aboutMe to the profile model.

MilesDev91 committed:
Merge branch 'manage-bookings-integration' into profile-details-page-integration

MilesDev91 committed:
Load profile with params

Passed params into profile details page, which we then use to retrieve a profile from the server via the API. Removed the mock profile.

Added payRate and location to the profile model. They are present on other branches as well.

MilesDev91 committed:
Profile photo

Added a way to pass down the profile photo url.

Set up the bottom of the details area for extra images once implemented.

Page displays differently if profile wasn't found

MilesDev91 committed:
Update ProfileDetails.tsx

Cleanup

MilesDev91 committed:
Requests from request from

Implemented sending requests from the request form.

Requester id no longer needs to be passed to create request, as we can find if with the user id.

Some cleanup

MilesDev91 commented:
I had to merge Bookings integration into this branch as there was a ton of Request related code I needed access to to complete this ticket.

sundayezeilo commented on server/controllers/profile.js:
```diff
@@ -38,4 +38,21 @@ exports.loadProfile = asyncHandler(async (req, res, next) => {
\ No newline at end of file
+});
+
+// @route GET /profile/load/:profileId
+// @route Get user profile data based on id
+// @access Public
```
:+1:

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+const asyncHandler = require("express-async-handler");
+const User = require("../models/User");
+
+// @route POST /requests
+// @desc Create new request
+// @access Public
```
Good documentation

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+  const { sitter, startDate, endDate } = req.body
+  const request = await Request.create({ requester: profileId, sitter, startDate, endDate });
+
+  if (request) {
+    res.status(200).json({
+      success: {
+        request: {
+          ...request
+        }
+      }
+    })
+  }
```
A failure message should also be returned to the client on failure

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+  }
+});
+
+// @route PUT /requests/edit/:requestId
+// @desc edit request
+// @access Public
```
:+1:

sundayezeilo reviewed (commented):
Nice integration :+1:

MilesDev91 committed:
Update request.js

MilesDev91 committed:
Fixes

Changed the controller to only allow editing of a request status that the user owns.

Simplified the retrieval of requester profiles for the sitter and fixed the front end code to match.

rajivtitus commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -4,95 +4,78 @@ import useStyles from './useStyles';
+    if (!mounted) {
+      setMounted(true);
+      getRequests().then((res) => {
+        console.log(res);
```
Please remove logs to the console

MilesDev91 replied:
Woops

rajivtitus reviewed (commented):
No major comments for this PR (I may have a couple of minor ones on second review)
Let's make the minor changes in Profile Details UI and update this PR with the changes and also with the latest remote main 👍🏽

MilesDev91 committed:
Merge branch 'profile-details-page-fe' into profile-details-page-integration

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -4,20 +4,33 @@ import AvatarDisplay from '../../components/AvatarDisplay/AvatarDisplay';
+      setIsMounted(true);
+      getProfile(profileId).then((res) => {
+        if (!res.error) {
+          setProfile(res.success.profile);
```
Let's update the Snackbar here in case we fail to fetch the sitter's profile

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,67 @@
+exports.getUserRequests = asyncHandler(async (req, res, next) => {
+  const profile = await Profile.findOne({'userId': req.user.id})
+  if (profile.isSitter) {
+    const requests = await Request.where('sitter', profile._id).populate('requester');
```
👍🏽

rajivtitus commented on client/src/pages/ProfileDetails/ProfileDetails.tsx:
```diff
@@ -50,47 +63,69 @@ export default function ProfileDetails(): JSX.Element {
           <Box m={windowWidth < 600 ? '1rem 0' : '5rem 0 1rem 0'} textAlign="center">
             <Typography fontWeight="bold" component="h1" fontSize="1.4rem">
-              {mockProfile.name}
+              {profile ? profile.name : 'No profile found'}
```
Why don't we consider rendering the entire Profile Details card only when a profile exists? In no profile is found we can just render large text - 'No Profile Found' (rather than displaying the cover photo and profile card) 😄

MilesDev91 replied:
That is much better than what I did. Thanks

rajivtitus reviewed (approved):
Nice work! Just a comment to address before merging 👍🏽

MilesDev91 committed:
Update ProfileDetails.tsx

Snackbar updated with error message on profile retrieval failure.

Wrapped entire profile card content in a profile check instead of individual components.

MilesDev91 committed:
Merge branch 'profile-details-page-fe' into profile-details-page-integration

MilesDev91 committed:
Bug fixes

Fixed a bug in a snackbar message.

Fixed id comparison in request controller.
